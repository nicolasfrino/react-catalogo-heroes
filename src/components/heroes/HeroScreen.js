import React, { useMemo} from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';
//import batman from '../../assets/heroes/dc-batman.jpg' //contenido estatico

//const pathImages = '../../assets/heroes/';
//const heroImages = require.context('../../assets/heroes', true); //context web pack

export const HeroScreen = ( {history} ) => {

    const {heroeId} =  useParams ();
    //console.log('heroeId', heroeId)

    //const hero = getHeroById(heroeId);
    const hero = useMemo(() => getHeroById(heroeId), [heroeId ]);


    //console.log('hero', hero)

    if (!hero) {
        return <Redirect to="/"/>
    }

    const handleReturn = () => {
        //console.log( 'history', history);

        if (history.length <= 2)
        {
            history.push('/');
        }else{
            history.goBack();
        }



    };

    const {
        superhero			,
        publisher			,
        alter_ego			,
        first_appearance	,
        characters			
    } = hero;

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img    
                        src={  `../../assets/heroes/${ heroeId }.jpg` }  //desde public/asset
                        //src={  `./../assets/heroes/${ heroeId }.jpg`.default }  //desde public/asset
                        //src={batman} //contenido estatico
                        //src={  (heroImages(`./${ heroeId }.jpg`).default) } //context web pack
                        alt={superhero}
                        className="img-thumbnail  animate__animated animate__fadeInLeft"
                />
            </div>

            <div className="col-8 animate__animated animate__fadeIn">
                <h3> {superhero} </h3>
                <ul className="list-group list-group-flush">

                    <li className="list-group-item"> <b> Alter Ego: </b>{ alter_ego } </li>
                    <li className="list-group-item"> <b> Publisher: </b>{ publisher } </li>
                    <li className="list-group-item"> <b> First appearance: </b>{ first_appearance } </li>

                </ul>

                <h5> Characters </h5>
                <p>  { characters }   </p>

                <button className="btn btn-outline-info"
                        onClick={handleReturn}
                >
                    Return
                </button>

            </div>

        </div>
    )
}
