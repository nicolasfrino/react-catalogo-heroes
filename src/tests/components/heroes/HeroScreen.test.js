import { mount } from "enzyme"
import React from "react"
import { MemoryRouter, Route } from "react-router-dom"
import { HeroScreen } from "../../../components/heroes/HeroScreen"

describe('HeroScreen', () => {


    test('debe de mostar el redirect si no hay argumento en la URL', () => {

        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount (
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={ historyMock } />
            </MemoryRouter>
        )

        expect (wrapper).toMatchSnapshot();
        expect (wrapper.find('Redirect').exists()).toBe(true);

    })
    
    
    test('debe de mostar un hero si el parametro existe y se encuentra', () => {

        const wrapper = mount (
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route exact path="/hero/:heroeId" component= { HeroScreen } />
            </MemoryRouter>
        )

        expect (wrapper.find('.row').exists()).toBe(true);

    })

    test('debe regresar a la pantalla anterior con push', () => {

        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount (
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route 
                    exact path="/hero/:heroeId" 
                    component= { (props) => <HeroScreen history = { historyMock } /> }/>
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')();
        expect (historyMock.push).toHaveBeenCalledWith('/');
        expect (historyMock.goBack).not.toHaveBeenCalled();

    })

    test('debe regresar a la pantalla anterior con goBack', () => {

        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount (
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route 
                    exact path="/hero/:heroeId" 
                    component= { (props) => <HeroScreen history = { historyMock } /> }/>
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')();
        expect (historyMock.goBack).toHaveBeenCalledWith();
        expect (historyMock.goBack).toHaveBeenCalledTimes(1);

        expect (historyMock.push).not.toHaveBeenCalled();
        expect (historyMock.push).toHaveBeenCalledTimes(0);

    })


    test('debe regresar a la pantalla de redirect si el hero no existe', () => {

        const historyMock = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn()
        }

        const wrapper = mount (
            <MemoryRouter initialEntries={['/hero/marvel-spider1234']}>
                <Route 
                    exact path="/hero/:heroeId" 
                    component= { (props) => <HeroScreen history = { historyMock } /> }/>
            </MemoryRouter>
        )

        expect (wrapper.text()).toBe('');

    })

})
