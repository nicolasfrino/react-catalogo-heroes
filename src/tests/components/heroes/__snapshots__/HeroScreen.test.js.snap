// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroScreen debe de mostar el redirect si no hay argumento en la URL 1`] = `""`;
